lib/world/provider.rb

```diff
@@ -0,0 +1,114 @@
+module Synthea
+  class Provider
+    cattr_accessor :provider_list # Array - all providers that are imported
+    cattr_accessor :services # Hash of services to hospitals that provide them
+
+    attr_accessor :attributes # Hash
+    attr_accessor :utilization # Hash
+    attr_accessor :services_provided # Array - services include ambulatory, emergency, and inpatient
+
+    def initialize(properties, coordinates)
+      @attributes = properties
+      @attributes[:coordinates] = coordinates
+
+      # labs = lipid_panel, basic_metabolic_panel, electrolytes_panel
+      # diabetes_labs = lipid_panel
+      @utilization = { encounters: 0, procedures: 0, labs: 0, diabetes_labs: 0, prescriptions: 0 }
+
+      Provider.provider_list.push(self)
+
+      @services_provided = []
+      services_split = properties['services_provided'].split
+      services_split.each do |service|
+        # add service to list of services provided by provider
+        @services_provided.push(service.to_sym)
+
+        # add provider to hash of services
+        if Provider.services.key?(service.to_sym)
+          Provider.services[service.to_sym].push(self)
+        else
+          Provider.services[service.to_sym] = [self]
+        end
+      end
+    end
+
+    # rubocop:disable Style/ClassVars
+    # from module.rb
+    def self.provider_list
+      @@provider_list ||= []
+    end
+
+    # from module.rb
+    def self.services
+      @@services ||= {}
+    end
+
+    def self.find_closest_service(entity, service)
+      if service == 'outpatient' || service == :outpatient
+        service = 'ambulatory'
+      end
+      # if service is provided by default hospital, go there
+      return entity.hospital if entity.hospital.service?(service.to_sym)
+
+      person_location = entity.attributes[:coordinates_address].to_coordinates
+      person_point = GeoRuby::SimpleFeatures::Point.from_x_y(person_location[0], person_location[1])
+      closest_distance = 100_000_000
+      closest_service = entity.hospital
+
+      # if service is nil or not supproted by simulation, patient goes to default hospital
+      if Provider.services.key?(service.to_sym)
+        Provider.services[service.to_sym].each do |h|
+          service_location = h.attributes[:coordinates]
```

Comment:
Rather than looping through all providers at every call, an alternative approach could be a voronoi diagram (or multiple diagrams, 1 per service/type) generated once when the app starts. Then when you look up "closest provider" it's more like O(1) than O(N). Can you run a quick test to see how much this impacts performance? Run 100 patients a few times with the latest from master and a few times with this branch and record the time it takes?

Just a note - I do see this library to create voronoi diagrams but it uses C extensions so I don't think we can use it. https://github.com/abscondment/rubyvor